zne/noise_amplification/folding_amplifier/glorious_folding_amplifier.py

```diff
@@ -10,21 +10,27 @@
 # copyright notice, and modified files need to carry a notice indicating
 # that they have been altered from the originals.
 
-"""Initial Function implmention. Will be evolved and documented more in depth"""
+"""Glorious DAG Folding Noise Amplification (Temporary)"""
+import copy
+
 from qiskit.circuit.library import Barrier
+from qiskit.converters import circuit_to_dag, dag_to_circuit
 from qiskit.dagcircuit import DAGCircuit
 
 
-def apply_folding(noisy_dag: DAGCircuit, original_dag: DAGCircuit, copies: int = 2) -> DAGCircuit:
-    """Initial function for now, will be evolved and documented more in depth"""
+def _inverse_dag(dag_to_inverse: DAGCircuit) -> DAGCircuit:
+    """Inverts dag circuit"""
+    dag_to_inverse = dag_to_circuit(dag_to_inverse).inverse()
+    return circuit_to_dag(dag_to_inverse)
 
-    for _ in range(copies):
-        noisy_dag.compose(original_dag, inplace=True)
+
+def apply_folding(original_dag: DAGCircuit, num_foldings: int) -> DAGCircuit:
+    """Build dag circuit by composing copies of an input dag circuit."""
+    noisy_dag = copy.deepcopy(original_dag)
+    inverse_dag = _inverse_dag(original_dag)
```

Comment:
Very good!! See if you can do the same with barriers so that you don't need to construct a new one every time. That should safe memory for the most part, instead of time! 😉